File: MyGame/app/src/main/java/com/example/mygame/domain/bullet/Bullet.kt
package com.example.mygame.domain.bullet

import com.example.mygame.domain.IGameObject
import com.example.mygame.domain.IMoveable
import com.example.mygame.domain.IVisitor

class Bullet(
    override var x: Float,
    override var y: Float,
    private val angle: Float
) : IMoveable, IGameObject {
    private var speedY = 0f

    override var isDisappeared = false

    override val left: Float
        get() = x - RADIUS / 2
    override val right: Float
        get() = x + RADIUS / 2
    override val top: Float
        get() = y - RADIUS / 2
    override val bottom: Float
        get() = y + RADIUS / 2



    fun shoot() {
        speedY = DEFAULT_ACCELERATION
    }

    override fun setPosition(startX: Float, startY: Float) {
        x = startX
        y = startY
    }

    override fun updatePosition(elapsedTime: Float) {
        y -= speedY * elapsedTime
        x += angle
    }

    override fun accept(visitor: IVisitor) {
        visitor.visit(this)
    }

    companion object {
        private const val RADIUS = 50f
        private const val DEFAULT_ACCELERATION = 2200f
    }
}

File: GameServer/src/Model/Bullet.php
<?php

namespace App\Model;

use App\Collision\IVisitor;

class Bullet implements IMoveable, IGameObject
{
    private const DEFAULT_ACCELERATION = 2200.0;
    private const RADIUS = 50.0;

    private $x;
    private $y;
    private $angle;
    private $isDisappeared = false;
    private $speedY = 0;

    public function __construct(float $x, float $y, float $angle)
    {
        $this->x = $x;
        $this->y = $y;
        $this->angle = $angle;
    }

    public function shoot(): void
    {
        $this->speedY = self::DEFAULT_ACCELERATION;
    }

    public function setPosition(float $startX, float $startY): void
    {
        $this->x = $startX;
        $this->y = $startY;
    }

    public function updatePosition(float $elapsedTime): void
    {
        $this->y -= $this->speedY * $elapsedTime;
        $this->x += $this->angle;
    }

    public function left(): float
    {
        return $this->x - self::RADIUS / 2;
    }

    public function right(): float
    {
        return $this->x + self::RADIUS / 2;
    }

    public function top(): float
    {
        return $this->y - self::RADIUS / 2;
    }

    public function bottom(): float
    {
        return $this->y + self::RADIUS / 2;
    }

    public function accept(IVisitor $visitor): void
    {
        $visitor->visitBullet($this);
    }

    public function getX(): float
    {
        return $this->x;
    }


    public function getY(): float
    {
        return $this->y;
    }


    public function getBottom(): float
    {
        return $this->y + self::RADIUS;
    }


    public function getIsDisappeared(): bool
    {
        return $this->isDisappeared;
    }


    public function getLeft(): float
    {
        return $this->x - self::RADIUS;
    }


    public function getRight(): float
    {
        return $this->x + self::RADIUS;
    }


    public function getTop(): float
    {
        return $this->y - self::RADIUS;
    }


    public function setIsDisappeared(bool $value): void
    {
        $this->isDisappeared = $value;
    }


    public function toArray(): array
    {
        return [];
    }
}
